Speech matching

PatrickNa opened the pull request:
This pull request adds updates to our speech recognition package. Further, the question matching package has been renewed. The name has been changed to 'speech matching' including also the 
categorization of 'commands'  (additional categories can be easily implemented). The last major contribution is the 'question answering' component. It receives from the speech matching node the question that has been recognized and lets the robot give and answer over the speakers.

This packages are still in their infancy, but the next steps are already planned and can be shared over interested team members easily.

Commit:
Switch from Julius to SpeechRecognition.

Instead of using Julius, the action now uses SpeechRecognition and the Google Service. In a future commit, also the offline solution with CMU Sphinx.

Commit:
Add offline speech recognition solution.

In case there is no internet connection available, the speech recognizer is going to use PocketSphinx from now on.

Commit:
Add offline speech recognition to mdr_speech_recognition.

With this commit also the mdr_speech_recognition package allows offline speech recognition using pocketsphinx. Same as for mdr_listen_action, also here google is preferred over pocketsphinx. In case there is a internet connection, the algorithm uses google otherwise pocketsphinx.

Commit:
Add speech matcher component.

The speech matcher is a revised version of the question matcher. It receives a string as input (recognized speech) and trys to match it to a sentence in the database. At the moment it uses the Levenshtein Distance to find the best match.

Commit:
Link speech recognition, speech matching and question answering.

This commit brings the previously commited components together by linking their publishers and subscribers. Additionally an own message type for the speech matching output has been created.

Commit:
Add launch files to speech_recognition, speech_matching and question_answering.

A simple launch file has been added to mdr_speech_recognition, mdr_speech_matching and mdr_question_answering. Beside, questions from the last German Open competition have been added and a fix in SpeechMatching has been made.

PatrickNa commented:
I don't agree with the issues from the Quality Review. Is it a must to resolve them all or can they be discussed?

argenos commented on mdr_speech/mdr_speech_matching/common/src/speech_matching/SpeechMatching.py:
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
```
The name of this file should be lower case and joined by underscores.

argenos commented on mdr_speech/mdr_question_answering/ros/src/mdr_question_answering/question_responder.py:
```diff
@@ -0,0 +1,47 @@
+def main():
+    rospy.init_node("question_responder")
+    question_responder = QuestionResponder()
+    sub = rospy.Subscriber("speech_matcher", MatchedSentence, question_responder.respond)
```
Why is this being initialized here and not in the constructor?

PatrickNa replied:
By initialized you mean the subscriber or the Voice message or both?

argenos replied:
why is it outside of the class main() instead of the constructor (`__init__`).

PatrickNa replied:
Good question! I change it in the evening...

argenos replied:
I think the node initialization and the subscriber belong in the constructor.

argenos replied:
And this probably belongs in the Answer action instead, so we can merge both

argenos reviewed (commented):
I have not looked at it thoroughly yet, but I removed some of the suggestions. The ones I think we shouldn't ignore I've commented on.

argenos commented:
This partially addresses https://github.com/b-it-bots/mas_domestic_robotics/issues/47, but leaves out the Answer action, correct?

PatrickNa commented:
Partially, however I would keep a node version of it, too.

PatrickNa commented:
Action answer I will add soon.

Commit:
Adjust lately created and renewed speech packages to coding convention.

This commit has several changes regarding syntax and logic. It is a first refactoring step, that is necessary at this point in order to fulfill the coding convention of the repository.

Commit:
Update the answer action.

The answer action now refers to the same answering logic as the package mdr_question_answering. It uses the same source for its answers and is adjusted to the new syntax in the answers.txt file. Additionally, the listen action matches the input sentence to a registered sentence in either questions.txt or commands.txt in the mdr_speech_matching package. It therefore uses the SpeechMatching class from the same package.

argenos commented on mdr_speech/mdr_speech_matching/ros/src/mdr_speech_matching/speech_matcher.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def main():
+    rospy.init_node("speech_matcher")
```
These also belong in the constructor.

argenos commented on mdr_speech/mdr_question_answering/ros/src/mdr_question_answering/question_responder.py:
```diff
@@ -0,0 +1,46 @@
+            rospy.logerr("No answer found!")
+        rospy.loginfo("This is not my type of sentence.")
+
+def main():
```
These belong in the constructor

argenos reviewed (dismissed):
I removed one of Codacy's suggestions, but I think we should keep the initialization of the nodes in the constructors instead of in the main function.

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_speech_actions/mdr_answer_action/ros/src/mdr_answer_action/action_states.py:
```diff
@@ -89,14 +99,19 @@ def execute(self, userdata):
-
-        if matching_line is None:
-            answer_file.close()
+        answer_dir = os.path.join(rospack.get_path("mdr_question_answering"), "ros/config/answers.txt")
```
The package and file name should be passed to the action as arguments (through the launcher and when the action server is created).

PatrickNa replied:
There is no launcher yet. I would be happy if this could be considered in an upcoming commit.

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_speech_actions/mdr_listen_action/README.md:
```diff
@@ -3,7 +3,7 @@ This package contains the action `listen`. It is used to be attentive to the env
 ## listen action - characteristics
 ### goal
-int32 listen_duration
+bool recognize_speech
```
What is the purpose of this parameter? I don't see it being used anywhere in the action implementation.

PatrickNa replied:
I don't know how to run it without specifying a goal. That's why defined something.

alex-mitrevski replied:
In the message definition, you can just leave the goal part empty (since the dashes delimit the different parts of the definition, there will be nothing above the first set of dashes).

alex-mitrevski commented on mdr_speech/mdr_question_answering/package.xml:
```diff
@@ -0,0 +1,29 @@
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="patrick.nagel@smail.inf.h-brs.de">Patrick Nagel</maintainer>
```
Please change the maintainer tag to
```
<maintainer email="robotics@inf.h-brs.de">MAS robotics</maintainer>
```
This eliminates the need for changing the tag when there are changes to the team.

Make the tags with yours and Roberto's name `author` tags.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_question_answering/ros/src/mdr_question_answering/question_responder.py:
```diff
@@ -0,0 +1,45 @@
+from std_msgs.msg import String
+from mdr_speech_matching.msg import MatchedSentence
+
+class QuestionResponder():
```
Inherit from `object` explicitly.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_speech_matching/common/src/speech_matching/speech_matching.py:
```diff
@@ -0,0 +1,80 @@
+- Double Metaphone (not implemented yet - probably fuzzy library)
+"""
+
+class SpeechMatching():
```
Inherit from `object` explicitly.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_speech_matching/ros/src/mdr_speech_matching/speech_matcher.py:
```diff
@@ -0,0 +1,39 @@
+from std_msgs.msg import String
+from mdr_speech_matching.msg import MatchedSentence
+
+class SpeechMatcher:
```
Inherit from `object` explicitly.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_speech_recognition/ros/src/mdr_speech_recognition/speech_recognizer.py:
```diff
@@ -0,0 +1,71 @@
+from std_msgs.msg import String
+import speech_recognition as sr
+
+class SpeechRecognizer:
```
Inherit from `object` explicitly.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_speech_matching/package.xml:
```diff
@@ -0,0 +1,28 @@
+  <version>0.0.1</version>
+  <description>The mdr_speech_matching package</description>
+
+  <maintainer email="patrick.nagel@smail.inf.h-brs.de">Patrick Nagel</maintainer>
```
Same comment as for `mdr_question_answering`.

PatrickNa replied:
done

alex-mitrevski commented on mdr_speech/mdr_speech_recognition/ros/src/mdr_speech_recognition/speech_recognizer.py:
```diff
@@ -0,0 +1,71 @@
+    @staticmethod
+    def check_internet_connection():
+        try:
+            urllib2.urlopen("http://172.217.21.238", timeout=1)
```
Why is this IP address hardcoded?

PatrickNa replied:
https://stackoverflow.com/questions/3764291/checking-network-connection

The answer to your question is in the chosen answer. I'm not sure if what the guy refers to is true for us, but I thought, why risk it.

alex-mitrevski replied:
Why not use the snippet that works with `httplib` instead (also in the same post)? That one doesn't require specifying an IP address - it's thus more robust - and it works just fine (I've just tested).

PatrickNa replied:
done

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_speech_actions/mdr_listen_action/README.md:
```diff
@@ -3,7 +3,7 @@ This package contains the action `listen`. It is used to be attentive to the env
 ## listen action - characteristics
 ### goal
-int32 listen_duration
+bool recognize_speech
```
Can you please remove this from the README?

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_speech_actions/mdr_answer_action/package.xml:
```diff
@@ -5,7 +5,7 @@
   <description>The mdr_answer_action package</description>
   <maintainer email="patrick.nagel@smail.inf.h-brs.de">Patrick Nagel</maintainer>
```
Just as for the other packages, make this an `author` tag and use

```
<maintainer email="robotics@inf.h-brs.de">MAS robotics</maintainer>
```

for the maintainer tag.
